torch/lib/TensorLib/TensorImpl.h

```diff
@@ -0,0 +1,37 @@
+#pragma once
+
+#include <atomic>
+
+#include "TensorLib/Scalar.h"
+#include "TensorLib/Type.h"
+#include <iostream>
+namespace tlib {
+
+class Type;
+struct TensorImpl {
```

Comment:
We use a PIMPL pattern. In PyTorch, THPTensor is a python object, where we use reference counting to control when to free the underlying TH/THC Tensor. In TensorLib, the TensorImpl can be thought of as the underlying memory to be managed. The Tensor itself handles reference counts on TensorImpl, so we can have many Tensors for each TensorImpl. In the same vein, we forward function calls to the implementation, just as we do in PyTorch.